### PR TITLE
test: add test to validate the controller actions to keep the cluster data

### PR DESCRIPTION
**Pull request by hectorj2f**
**What this PR does / why we need it**:

It is a continuation of https://github.com/kubernetes-sigs/kubefed/pull/1201 which has been stale for a long period. I cherry picked their changes and added some tests. 

This test also cover other functions of the kubefedcluster controller.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/controller/kubefedcluster/controller_test.go [hunk 1]
```diff
@@ -17,19 +17,102 @@ limitations under the License.
...
 	"time"
 
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	//schema "k8s.io/apimachinery/pkg/runtime/schema"
```
Comment (irfanurrehman) [inline pkg/controller/kubefedcluster/controller_test.go:32]: Please remove the commented lines here and at other places below.

File: pkg/controller/kubefedcluster/controller_test.go [hunk 2]
```diff
@@ -38,71 +121,124 @@ func TestThresholdCheckedClusterStatus(t *testing.T) {
...
+		Expect(err).ToNot(HaveOccurred())
+
+		fedCluster.Spec.APIEndpoint = "https://my.example.com:80/path/to/newendpoint"
+		err = k8sClient.Update(ctx, fedCluster)
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(func() bool {
+			_, found = cc.clusterDataMap[kc.Name]
+			return found
+		}, 30*time.Second).Should(BeTrue())
```
Comment (irfanurrehman) [inline pkg/controller/kubefedcluster/controller_test.go:179]: Recommend: Verify that the data updated (API Endpoint) equals what is updated above?

**Comment by irfanurrehman** [comment]
@hectorj2f Thanks for doing this. The test itself looks quite allright. A few nits (see comments).

Additionally little bit of history for your reference:
We had the integration tests (the kind that your wrote here) earlier, where a local etcd and api server were used for individual tests; however we saw that the overall time taken to complete all tests against the PRs was crossing 1 hour. At some point we did fold all those integration tests into e2e to save time (currently the test run still takes close to 40+ mins).
The difference I am talking about is:
Test run with etcd and api-server in this PR: [7.507s](https://travis-ci.org/github/kubernetes-sigs/kubefed/builds/674631544#L3042) 
Pure unit test from an older build: [.021s](https://travis-ci.org/github/kubernetes-sigs/kubefed/builds/623483573#L2040)

Having said that, I do agree that the current test is not verifying the cluster state change and I think its ok to introduce the new test you have added, covering the missing piece as here (It would be worse in terms of time consumed to verify this scenario in e2e).
Can I also recommend that the current unit test be left behind as it is, and add the test that you have introduced separately under test/integration (If you think its easy enough to do)?
Again, really appreciate you doing this.

File: pkg/controller/kubefedcluster/controller_test.go (hunk 2 above)
Comment (hectorj2f) [reply pkg/controller/kubefedcluster/controller_test.go:179]: 👍

**Comment by hectorj2f** [comment]
@irfanurrehman I'd check how this can all be introduced without causing a lot of confusion or messing up the code structure. I certainly realized that there are many pieces of the kubefed's code that are not tested either because it escapes from the e2e radar or it is too complex to create mocks for all these tess.

**Comment by hectorj2f** [comment]
@irfanurrehman I addressed your comments. I didn't move the new test to `test/integration` due to the amount of additional changes that will imply.

**Comment by irfanurrehman** [comment]
Thanks @hectorj2f 
/lgtm

**Pull request merged**